File: ora_flutter/lib/features/auth/application/auth_repository.dart
import '../../../core/network/apps_script_client.dart';
import '../data/auth_api.dart';
import '../data/session_store.dart';
import '../domain/auth_models.dart';
import '../domain/auth_validation.dart';

typedef Clock = DateTime Function();

class AuthRepository {
  AuthRepository({required this.api, required this.sessionStore, Clock? clock})
    : clock = clock ?? DateTime.now;

  final AuthApi api;
  final SessionStore sessionStore;
  final Clock clock;

  Future<UserSession?> restoreSession() async {
    try {
      final session = await sessionStore.load();
      if (session == null) return null;
      if (session.isExpiredAt(clock()) ||
          session.status.toUpperCase() != 'ACTIVE') {
        await sessionStore.clear();
        return null;
      }
      return session;
    } on Object {
      await sessionStore.clear();
      return null;
    }
  }

  Future<AuthOutcome> login(String nikInput, String pin) async {
    final nikError = nikValidationError(nikInput);
    if (nikError != null) throw AuthFailure(nikError);
    final pinError = pinValidationError(pin);
    if (pinError != null) throw AuthFailure(pinError);

    try {
      final result = await api.login(nikInput.trim(), pin);
      if (!result.participant.isActive) {
        throw const AuthFailure('This account is inactive.');
      }
      final expiresAt = clock().add(Duration(seconds: result.expiresInSeconds));
      if (result.requiresNicknameActivation ||
          result.participant.nickname == null) {
        return ActivationRequired(
          PendingActivation(
            nik: result.participant.nik,
            divisionGuild: result.participant.divisionGuild,
            sessionToken: result.sessionToken,
            sessionExpiresAt: expiresAt,
          ),
        );
      }
      return await _authenticate(
        UserSession(
          sessionToken: result.sessionToken,
          nik: result.participant.nik,
          nickname: result.participant.nickname!,
          divisionGuild: result.participant.divisionGuild,
          status: result.participant.status,
          expiresAt: expiresAt,
        ),
      );
    } on BackendFailure catch (error) {
      throw AuthFailure(_userMessage(error));
    }
  }

  Future<AuthOutcome> activate(
    PendingActivation pending,
    String nicknameInput,
  ) async {
    final error = nicknameValidationError(nicknameInput);
    if (error != null) throw AuthFailure(error);
    if (!pending.sessionExpiresAt.isAfter(clock())) {
      await sessionStore.clear();
      throw const AuthFailure(
        'Session expired. Please login again.',
        sessionInvalid: true,
      );
    }
    final nickname = canonicalNickname(nicknameInput);
    try {
      final participant = await api.activateNickname(
        pending.sessionToken,
        nickname,
      );
      if (!participant.isActive) {
        throw const AuthFailure('This account is inactive.');
      }
      return await _authenticate(
        UserSession(
          sessionToken: pending.sessionToken,
          nik: participant.nik,
          nickname: participant.nickname ?? nickname,
          divisionGuild: participant.divisionGuild,
          status: participant.status,
          expiresAt: pending.sessionExpiresAt,
        ),
      );
    } on BackendFailure catch (error) {
      throw AuthFailure(
        _userMessage(error),
        sessionInvalid: error.invalidatesSession,
      );
    }
  }

  Future<UserSession> updateNickname(
    UserSession current,
    String nicknameInput,
  ) async {
    final error = nicknameValidationError(nicknameInput);
    if (error != null) throw AuthFailure(error);
    if (current.isExpiredAt(clock())) {
      await sessionStore.clear();
      throw const AuthFailure(
        'Session expired. Please login again.',
        sessionInvalid: true,
      );
    }
    final nickname = canonicalNickname(nicknameInput);
    try {
      final participant = await api.updateNickname(
        current.sessionToken,
        nickname,
      );
      if (participant.nik != current.nik || !participant.isActive) {
        throw const AuthFailure('Unable to update nickname.');
      }
      final updated = current.copyWith(
        nickname: participant.nickname ?? nickname,
      );
      await sessionStore.save(updated);
      return updated;
    } on BackendFailure catch (error) {
      throw AuthFailure(
        _userMessage(error),
        sessionInvalid: error.invalidatesSession,
      );
    }
  }

  Future<Authenticated> _authenticate(UserSession session) async {
    await sessionStore.save(session);
    return Authenticated(session);
  }

  Future<void> logout() => sessionStore.clear();

  String _userMessage(BackendFailure error) => switch (error.code) {
    'INVALID_CREDENTIALS' => 'NIK or PIN is incorrect.',
    'ACCOUNT_INACTIVE' => 'This account is inactive.',
    'NICKNAME_TAKEN' => 'Nickname is already in use.',
    'INVALID_NICKNAME' =>
      error.message.isEmpty ? 'Nickname is invalid.' : error.message,
    'NICKNAME_ALREADY_ACTIVATED' =>
      'Nickname has already been activated. Please login again.',
    'SESSION_EXPIRED' ||
    'UNAUTHORIZED' => 'Session expired. Please login again.',
    _ => 'Unable to connect to ORA. Check your connection and try again.',
  };
}

class AuthFailure implements Exception {
  const AuthFailure(this.message, {this.sessionInvalid = false});
  final String message;
  final bool sessionInvalid;
}

File: ORA/app/src/main/java/com/otorunners/ora/auth/AuthRepository.kt
package com.otorunners.ora.auth

import com.otorunners.ora.backend.AppsScriptBackendApi
import com.otorunners.ora.backend.BackendException
import com.otorunners.ora.backend.OraBackendApi
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class AuthRepository(
    private val authStore: AuthStore,
    private val backendApi: OraBackendApi = AppsScriptBackendApi(),
    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
    private val nowProvider: () -> Long = { System.currentTimeMillis() }
) {
    fun restoreSession(): UserSession? = authStore.loadSession()

    suspend fun login(nikInput: String, pin: String): AuthOutcome = withContext(ioDispatcher) {
        val nik = nikInput.trim()
        if (nik.isEmpty()) return@withContext AuthOutcome.Error("Enter your NIK.")
        pinValidationError(pin)?.let { return@withContext AuthOutcome.Error(it) }

        try {
            val result = backendApi.login(nik, pin)
            val expiresAtMillis = nowProvider() + result.expiresInSeconds * 1_000L
            if (result.requiresNicknameActivation || result.participant.nickname.isNullOrBlank()) {
                AuthOutcome.ActivationRequired(
                    PendingActivation(
                        nik = result.participant.nik,
                        divisionGuild = result.participant.divisionGuild,
                        sessionToken = result.sessionToken,
                        sessionExpiresAtMillis = expiresAtMillis
                    )
                )
            } else {
                authenticate(
                    UserSession(
                        nik = result.participant.nik,
                        nickname = result.participant.nickname,
                        divisionGuild = result.participant.divisionGuild,
                        active = result.participant.active,
                        sessionToken = result.sessionToken,
                        sessionExpiresAtMillis = expiresAtMillis
                    )
                )
            }
        } catch (error: Exception) {
            AuthOutcome.Error(error.toUserMessage())
        }
    }

    suspend fun activate(pending: PendingActivation, nicknameInput: String): AuthOutcome =
        withContext(ioDispatcher) {
            nicknameValidationError(nicknameInput)?.let {
                return@withContext AuthOutcome.Error(it)
            }
            val nickname = canonicalNickname(nicknameInput)
            try {
                val participant = backendApi.activateNickname(pending.sessionToken, nickname)
                authenticate(
                    UserSession(
                        nik = participant.nik,
                        nickname = participant.nickname ?: nickname,
                        divisionGuild = participant.divisionGuild,
                        active = participant.active,
                        sessionToken = pending.sessionToken,
                        sessionExpiresAtMillis = pending.sessionExpiresAtMillis
                    )
                )
            } catch (error: Exception) {
                AuthOutcome.Error(error.toUserMessage())
            }
        }

    fun renameNickname(currentUser: UserSession, nicknameInput: String): AuthOutcome {
        nicknameValidationError(nicknameInput)?.let { return AuthOutcome.Error(it) }

        val nickname = canonicalNickname(nicknameInput)
        if (authStore.isNicknameTaken(nickname, currentUser.nik)) {
            return AuthOutcome.Error("Nickname is already in use.")
        }

        return authenticate(currentUser.copy(nickname = nickname))
    }

    fun logout() {
        authStore.clearSession()
    }

    private fun authenticate(user: UserSession): AuthOutcome.Authenticated {
        authStore.saveActivatedProfile(user)
        authStore.saveSession(user)
        return AuthOutcome.Authenticated(user)
    }

    private fun Exception.toUserMessage(): String {
        return when ((this as? BackendException)?.code) {
            "INVALID_CREDENTIALS" -> "NIK or PIN is incorrect."
            "ACCOUNT_INACTIVE" -> "This account is inactive."
            "NICKNAME_TAKEN" -> "Nickname is already in use."
            "INVALID_NICKNAME" -> message ?: "Nickname is invalid."
            "SESSION_EXPIRED", "UNAUTHORIZED" -> "Session expired. Please login again."
            else -> "Unable to connect to ORA. Check your connection and try again."
        }
    }
}
